core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -443,9 +466,27 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorIdle(executorId: String): Unit = synchronized {
     if (executorIds.contains(executorId)) {
       if (!removeTimes.contains(executorId) && !executorsPendingToRemove.contains(executorId)) {
+
+        val hasCachedBlocks =
+          executorsWithCachedBlocks.contains(executorId) ||
+            executorEndpoints.get(executorId).exists(_.askWithRetry[Boolean](HasCachedBlocks))
+
+        if (hasCachedBlocks) executorsWithCachedBlocks += executorId
```

Comment:
i donot know why we should have an RPC for every executor? I have read codes of BlockManager carefully and find that blocks except BroadcastBlockId can tell BlockMaster immediately when cached or unpersist. maybe  i have some mistakes?
if we cannot read blockStatus from BlockManager, i think executor can report hasCachedBlocks to Driver when Heartbeat. so we donot need to have an RPC for every executor because that is very expensive.
@andrewor14 @harishreedharan